RustPython

**Project description**
A Python-3 (CPython >= 3.9.0) Interpreter written in Rust 

**Metadata**

* homepage URL: https://rustpython.github.io
* source URL: https://github.com/RustPython/RustPython
* license: mit 
* platforms: unix, linux, darwin

I tried to package it here: https://github.com/prusnak/nixpkgs/tree/rustpython

but run into compilation failure:

```
error[E0308]: mismatched types
  --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/bhcomp.rs:62:24
   |
62 |     let bytes = bits / Limb::BITS;
   |                        ^^^^^^^^^^ expected `usize`, found `u32`

error[E0277]: cannot divide `usize` by `u32`
  --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/bhcomp.rs:62:22
   |
62 |     let bytes = bits / Limb::BITS;
   |                      ^ no implementation for `usize / u32`
   |
   = help: the trait `Div<u32>` is not implemented for `usize`

error[E0308]: mismatched types
   --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/bigcomp.rs:158:55
    |
158 |     let nlz = den.leading_zeros().wrapping_sub(wlz) & (u32::BITS - 1);
    |                                                       ^^^^^^^^^^^^^^^ expected `usize`, found `u32`

error[E0277]: no implementation for `usize & u32`
   --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/bigcomp.rs:158:53
    |
158 |     let nlz = den.leading_zeros().wrapping_sub(wlz) & (u32::BITS - 1);
    |                                                     ^ no implementation for `usize & u32`
    |
    = help: the trait `BitAnd<u32>` is not implemented for `usize`

error[E0308]: mismatched types
   --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/bigcomp.rs:176:40
    |
176 |         let (q, r) = shift.ceil_divmod(Limb::BITS);
    |                                        ^^^^^^^^^^ expected `usize`, found `u32`
    |
help: you can convert a `u32` to a `usize` and panic if the converted value doesn't fit
    |
176 |         let (q, r) = shift.ceil_divmod(Limb::BITS.try_into().unwrap());
    |                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:2071:27
     |
2071 |     let rs = Limb::BITS - s;
     |                           ^ expected `u32`, found `usize`

error[E0277]: cannot subtract `usize` from `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:2071:25
     |
2071 |     let rs = Limb::BITS - s;
     |                         ^ no implementation for `u32 - usize`
     |
     = help: the trait `Sub<usize>` is not implemented for `u32`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1049:42
     |
1049 |     let mut count = index.saturating_mul(Limb::BITS);
     |                                          ^^^^^^^^^^ expected `usize`, found `u32`
     |
help: you can convert a `u32` to a `usize` and panic if the converted value doesn't fit
     |
1049 |     let mut count = index.saturating_mul(Limb::BITS.try_into().unwrap());
     |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1064:28
     |
1064 |     Limb::BITS.checked_mul(x.len())
     |                            ^^^^^^^ expected `u32`, found `usize`
     |
help: you can convert a `usize` to a `u32` and panic if the converted value doesn't fit
     |
1064 |     Limb::BITS.checked_mul(x.len().try_into().unwrap())
     |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1065:22
     |
1065 |         .map(|v| v - nlz)
     |                      ^^^ expected `u32`, found `usize`

error[E0277]: cannot subtract `usize` from `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1065:20
     |
1065 |         .map(|v| v - nlz)
     |                    ^ no implementation for `u32 - usize`
     |
     = help: the trait `Sub<usize>` is not implemented for `u32`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1066:20
     |
1066 |         .unwrap_or(usize::max_value())
     |                    ^^^^^^^^^^^^^^^^^^ expected `u32`, found `usize`
     |
help: you can convert a `usize` to a `u32` and panic if the converted value doesn't fit
     |
1066 |         .unwrap_or(usize::max_value().try_into().unwrap())
     |                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1064:5
     |
1060 |   pub fn bit_length(x: &[Limb]) -> usize {
     |                                    ----- expected `usize` because of return type
...
1064 | /     Limb::BITS.checked_mul(x.len())
1065 | |         .map(|v| v - nlz)
1066 | |         .unwrap_or(usize::max_value())
     | |______________________________________^ expected `usize`, found `u32`
     |
help: you can convert a `u32` to a `usize` and panic if the converted value doesn't fit
     |
1064 ~     Limb::BITS.checked_mul(x.len())
1065 +         .map(|v| v - nlz)
1066 +         .unwrap_or(usize::max_value()).try_into().unwrap()
     |

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1091:23
     |
1091 |     debug_assert!(n < bits && n != 0);
     |                       ^^^^ expected `usize`, found `u32`
     |
help: you can convert a `u32` to a `usize` and panic if the converted value doesn't fit
     |
1091 |     debug_assert!(n < bits.try_into().unwrap() && n != 0);
     |                       ~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1098:25
     |
1098 |     let lshift = bits - n;
     |                         ^ expected `u32`, found `usize`

error[E0277]: cannot subtract `usize` from `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1098:23
     |
1098 |     let lshift = bits - n;
     |                       ^ no implementation for `u32 - usize`
     |
     = help: the trait `Sub<usize>` is not implemented for `u32`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1140:19
     |
1140 |     let rem = n % bits;
     |                   ^^^^ expected `usize`, found `u32`

error[E0277]: cannot mod `usize` by `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1140:17
     |
1140 |     let rem = n % bits;
     |                 ^ no implementation for `usize % u32`
     |
     = help: the trait `Rem<u32>` is not implemented for `usize`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1141:19
     |
1141 |     let div = n / bits;
     |                   ^^^^ expected `usize`, found `u32`

error[E0277]: cannot divide `usize` by `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1141:17
     |
1141 |     let div = n / bits;
     |                 ^ no implementation for `usize / u32`
     |
     = help: the trait `Div<u32>` is not implemented for `usize`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1197:23
     |
1197 |     debug_assert!(n < bits);
     |                       ^^^^ expected `usize`, found `u32`
     |
help: you can convert a `u32` to a `usize` and panic if the converted value doesn't fit
     |
1197 |     debug_assert!(n < bits.try_into().unwrap());
     |                       ~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1207:25
     |
1207 |     let rshift = bits - n;
     |                         ^ expected `u32`, found `usize`

error[E0277]: cannot subtract `usize` from `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1207:23
     |
1207 |     let rshift = bits - n;
     |                       ^ no implementation for `u32 - usize`
     |
     = help: the trait `Sub<usize>` is not implemented for `u32`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1259:19
     |
1259 |     let rem = n % bits;
     |                   ^^^^ expected `usize`, found `u32`

error[E0277]: cannot mod `usize` by `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1259:17
     |
1259 |     let rem = n % bits;
     |                 ^ no implementation for `usize % u32`
     |
     = help: the trait `Rem<u32>` is not implemented for `usize`

error[E0308]: mismatched types
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1260:19
     |
1260 |     let div = n / bits;
     |                   ^^^^ expected `usize`, found `u32`

error[E0277]: cannot divide `usize` by `u32`
    --> /private/tmp/nix-build-rustpython-0.1.2.drv-0/rustpython-0.1.2-vendor.tar.gz/lexical-core/src/atof/algorithm/math.rs:1260:17
     |
1260 |     let div = n / bits;
     |                 ^ no implementation for `usize / u32`
     |
     = help: the trait `Div<u32>` is not implemented for `usize`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `lexical-core` due to 27 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
error: builder for '/nix/store/alk3l6qdfddc8zlwljxmgz9f4rnynx2n-rustpython-0.1.2.drv' failed with exit code 101
```

--- a/pkgs/development/interpreters/python/default.nix
+++ b/pkgs/development/interpreters/python/default.nix
@@ -285,4 +285,8 @@ in {
     inherit passthruFun;
   };
 
+  rustpython = callPackage ./rustpython/default.nix {
+    inherit (darwin.apple_sdk.frameworks) SystemConfiguration;
+  };
+
 })

--- a/pkgs/development/interpreters/python/rustpython/default.nix
+++ b/pkgs/development/interpreters/python/rustpython/default.nix
@@ -1,0 +1,44 @@
+{ lib
+, stdenv
+, rustPlatform
+, fetchFromGitHub
+, SystemConfiguration
+, python3
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "rustpython";
+  version = "unstable-2021-12-09";
+
+  src = fetchFromGitHub {
+    owner = "RustPython";
+    repo = "RustPython";
+    rev = "db3b3127df34ff5dd569301aa36ed71ae5624e4e";
+    sha256 = "sha256-YwGfXs3A5L/18mHnnWubPU3Y8EI9uU3keJ2HJnnTwv0=";
+  };
+
+  cargoHash = "sha256-T85kiPG80oZ4mwpb8Ag40wDHKx2Aens+gM7NGXan5lM=";
+
+  # freeze the stdlib into the rustpython binary
+  cargoBuildFlags = "--features=freeze-stdlib";
+
+  buildInputs = lib.optionals stdenv.isDarwin [ SystemConfiguration ];
+
+  checkInputs = [ python3 ];
+
+  meta = with lib; {
+    description = "Python 3 interpreter in written Rust";
+    homepage = "https://rustpython.github.io";
+    license = licenses.mit;
+    maintainers = with maintainers; [ prusnak ];
+
+    # TODO: Remove once nixpkgs uses newer SDKs that supports '*at' functions.
+    # Probably macOS SDK 10.13 or later. Check the current version in
+    # .../os-specific/darwin/apple-sdk/default.nix
+    #
+    # From the build logs:
+    #
+    # > Undefined symbols for architecture x86_64: "_utimensat"
+    broken = stdenv.isDarwin && stdenv.isx86_64;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -13562,7 +13562,7 @@ with pkgs;
   python3Packages = python3.pkgs;
 
   pythonInterpreters = callPackage ./../development/interpreters/python { };
-  inherit (pythonInterpreters) python27 python37 python38 python39 python310 python3Minimal pypy27 pypy38;
+  inherit (pythonInterpreters) python27 python37 python38 python39 python310 python3Minimal pypy27 pypy38 rustpython;
 
   # Python package sets.
   python27Packages = python27.pkgs;

